### DISCUSSION TITLE
What does "disable Automatic Static Optimization in pages without Static Generation" mean?

### DISCUSSION THREAD

**Discussion by cbdeveloper** [vercel/next.js#40082] [Help] [answered]
In the caveats of implementing `getInitialProps` on custom `App`, we get:

https://nextjs.org/docs/advanced-features/custom-app#caveats

![image](https://user-images.githubusercontent.com/43407798/187372393-d11a386e-2555-4276-a54c-55ea88e9f8cf.png)

DIsabling _Automatic Static Optimization_, means that my page will be server side rendered, instead of statically rendered during build, right?

But it says that it'll only happen for pages _without Static Generation_.

That's where it loses me a little.

This is what it says about _Static Generation_

![image](https://user-images.githubusercontent.com/43407798/187372924-65eadaa6-b146-4223-ba72-c6b8bff90b6e.png)

Am I right assuming the following:

- I implement `getInitialProps` on my custom `App`
- Assume `PageA` doesn't implement `getServerSideProps` and `getInitialProps`
- That would disable static generation on my `PageA` (because of the caveat).
- But if `PageA` explicitly implements `getStaticProps`
- It will still be statically built (pre-rendered during build). So it basically reverts the caveat?

Is that correct?

**Comment by icyJoseph** [comment, marked as answer]
If you define `getInitialProps` in your `_app` component then:

- When Next.js sees a page with `getStaticProps`, it'll run it at build time (`_app.getInitialProps` + `getStaticProps`)
   -  This means that if you use `req` for example, the build process might fail depending on what you want to do with `req`, and you'll definitely see variation of props/context and other global data applied to your app.
- Every other page is rendered as requested. 

So your assumption is correct, `getStaticProps` is a way to tell Next.js that you are guarantee that the page can be built during `next build`.

When you build with `_app.getInitialProps` present, you'll see this warning:

```shell
Warning: You have opted-out of Automatic Static Optimization due to `getInitialProps` in `pages/_app`. This does not opt-out pages with `getStaticProps`
Read more: https://nextjs.org/docs/messages/opt-out-auto-static-optimization
```

And your summary will detail what has been done, and how each page will behave at runtime:

```shell
Page                                       Size     First Load JS
┌ λ /                                      4.57 kB          84 kB
├   /_app                                  0 B            77.2 kB
├ λ /404                                   194 B          77.4 kB
├ ℇ /api/edge                              0 B            77.2 kB
├ λ /api/hello                             0 B            77.2 kB
└● /blog                                  319 B          79.7 kB

+ First Load JS shared by all              77.2 kB
  ├ chunks/framework-7dc8a65f4a0cda33.js   45.2 kB
  ├ chunks/main-05fed6726b331f18.js        30.7 kB
  ├ chunks/pages/_app-33a7e1fff22d6141.js  528 B
  ├ chunks/webpack-69bfa6990bb9e155.js     769 B
  └ css/27d177a30947857b.css               194 B

ƒ Middleware                               19.3 kB

ℇ  (Streaming)  server-side renders with streaming (uses React 18 SSR streaming or Server Components)
λ  (Server)     server-side renders at runtime (uses getInitialProps or getServerSideProps)
●  (SSG)        automatically generated as static HTML + JSON (uses getStaticProps)
```

In this case there's a `getInitialProps` function in `_app`, which is why, `/` is a `λ` type of page, whereas, because `/blog` implements `getStaticProps` it is labelled as ●.

The two biggest issues are access to `req`, `res`, headers, cookies and such, which will not be present at build time. And, then there's the problem that if the result of `getInitialProps` is used to initialise some global data, when you go to a page that was rendered at build time, you'll need to make sure that the input to the context provider is correct, in terms of your business logic. 

I think white label apps often run into issues with this, where the pages that look the same to all pages are pre-rendered, but they inject some generic global data to the context provider, and thus, navigating into them or away from them may trigger nasty bugs.

**Reply by cbdeveloper** [reply]
@icyJoseph thank you so much for this very detailed answer.

I get it now. So, basically, if I'm not using anything `request/response` specific, I should be ok, right?

Anyway, I'll try to avoid the `getInitialProps` on `_app`.

My plan with this was to inject some global data on `_app` level, that other static pages would need. I was looking into saving some DB requests by doing that. Assuming it would run only once during build.

Does `getInitialProps` on `_app` run once per project-build? Or does it run once per page-build? Do you know that?

**Reply by icyJoseph** [reply]
At build time it runs per page build. So it'll run for as many pages as you generate.

**Comment by dsteinbach-ep** [comment]
Is there a way to purposely opt out of static optimization at build time?

**Reply by andrewmartin** [reply]
Following up on this comment, I'm having some very odd behavior on CI systems (Vercel, Netlify BOTH). Would love to disable this altogether if possible.
